src/xdplwf/recv.c

```diff
@@ -890,54 +1106,122 @@ XdpGenericReceiveTxInspectPoll(
+    TraceEnter(
+        TRACE_GENERIC, "RxQueue=%p IfIndex=%u QueueId=%u TxInspect=%!BOOLEAN!",
+        RxQueue, Generic->IfIndex, RxQueue->QueueId, RxQueue->Flags.TxInspect);
+
+    KeAcquireSpinLock(&RxQueue->EcLock, &OldIrql);
+    if (!RxQueue->Flags.Paused) {
+        RxQueue->Flags.Paused = TRUE;
+        NeedPause = TRUE;
+    }
+    KeReleaseSpinLock(&RxQueue->EcLock, OldIrql);
 
-    if (NeedRestart) {
-        TraceVerbose(TRACE_GENERIC, "IfIndex=%u Requesting RX datapath attach", Generic->IfIndex);
-        XdpGenericRequestRestart(Generic);
+    if (NeedPause) {
+        ExWaitForRundownProtectionRelease(&RxQueue->NblRundown);
     }
 
-    Timeout.QuadPart =
-        -1 * RTL_MILLISEC_TO_100NANOSEC(GENERIC_DATAPATH_RESTART_TIMEOUT_MS);
-    KeWaitForSingleObject(
-        &Datapath->ReadyEvent, Executive, KernelMode, FALSE, &Timeout);
+    TraceExitSuccess(TRACE_GENERIC);
 }
 
 _IRQL_requires_max_(PASSIVE_LEVEL)
+_Requires_exclusive_lock_held_(&Generic->Lock)
 VOID
-XdpGenericDetachIfRx(
+XdpGenericRxPause(
+    _In_ XDP_LWF_GENERIC *Generic
+    )
+{
+    TraceEnter(TRACE_GENERIC, "IfIndex=%u", Generic->IfIndex);
+
+    LIST_ENTRY *Entry = Generic->Rx.Queues.Flink;
+    while (Entry != &Generic->Rx.Queues) {
+        XDP_LWF_GENERIC_RX_QUEUE *RxQueue =
+            CONTAINING_RECORD(Entry, XDP_LWF_GENERIC_RX_QUEUE, Link);
+        Entry = Entry->Flink;
+
+        XdpGenericRxPauseQueue(Generic, RxQueue);
+    }
+
+    TraceExitSuccess(TRACE_GENERIC);
+}
+
+static
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_Requires_exclusive_lock_held_(&Generic->Lock)
+VOID
+XdpGenericRxRestartQueue(
     _In_ XDP_LWF_GENERIC *Generic,
-    _In_ XDP_LWF_DATAPATH_BYPASS *Datapath
+    _In_ XDP_LWF_GENERIC_RX_QUEUE *RxQueue
     )
 {
-    BOOLEAN NeedRestart;
+    KIRQL OldIrql;
 
-    RtlAcquirePushLockExclusive(&Generic->Lock);
-    XdpGenericDereferenceDatapath(Generic, Datapath, &NeedRestart);
-    RtlReleasePushLockExclusive(&Generic->Lock);
+    TraceEnter(
+        TRACE_GENERIC, "RxQueue=%p IfIndex=%u QueueId=%u TxInspect=%!BOOLEAN!",
+        RxQueue, Generic->IfIndex, RxQueue->QueueId, RxQueue->Flags.TxInspect);
+
+    KeAcquireSpinLock(&RxQueue->EcLock, &OldIrql);
+    ASSERT(RxQueue->Flags.Paused);
+    RxQueue->Flags.Paused = FALSE;
+    KeReleaseSpinLock(&RxQueue->EcLock, OldIrql);
+
+    ExReInitializeRundownProtection(&RxQueue->NblRundown);
+
+    TraceExitSuccess(TRACE_GENERIC);
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_Requires_exclusive_lock_held_(&Generic->Lock)
+VOID
+XdpGenericRxRestart(
+    _In_ XDP_LWF_GENERIC *Generic,
+    _In_ UINT32 NewMtu
+    )
+{
+    LIST_ENTRY *Entry = Generic->Rx.Queues.Flink;
+
+    TraceEnter(TRACE_GENERIC, "IfIndex=%u NewMtu=%u", Generic->IfIndex, NewMtu);
+
+    //
+    // For RX, the interface MTU is currently not needed. To handle encap or
+    // offloads, though, we will eventually need to track the MTU.
+    //
+    UNREFERENCED_PARAMETER(NewMtu);
```

Comment:
Just curious, does this mean eventually you might even need to generate ICMP errors if things are too big to encap?

Reply:
XDP would never be in the business of generating ICMP errors. Whatever program (XDP or eBPF) would have to be responsible for checking the MTU before attempting encap if it wants to have protocol layer correctness. XDP is only responsible for driver correctness, and it is illegal to send a non-offload packet larger than MTU.

Reply:
And BTW, we will probably implement that MTU check within the XDP helper that allows packets to be retreated for encap. This is just a bit of plumbing to be standardized with the equivalent TX code path for now.